### PR TITLE
Implement .keys, .values, .iter for Map

### PR DESCRIPTION
**Pull request by 5225225**
One thing I'm not sure of is if we should be guaranteeing that the iterators return items in order of the key, like the backing BTree does. I'm playing it safe by sorting the items before the comparison and not saying we return in order of the key.

I don't have any use for that property, though.

### CODE REVIEW

File: src/map.rs [hunk 1]
```diff
@@ -344,6 +344,139 @@ impl<K: Ord, V: Val<A> + Default, A: Actor> Map<K, V, A> {
...
+    /// }
+    ///
+    /// items.sort();
+    ///
+    /// assert_eq!(items, &["bar", "baz", "foo"]);
+    /// ```
+    pub fn values(&self) -> impl IntoIterator<Item = ReadCtx<&V, A>> {
+        self.entries.values().map(move |v| ReadCtx {
+            add_clock: self.clock.clone(),
+            rm_clock: self.clock.clone(),
```
Comment (5225225) [inline src/map.rs:431]: Not sure if this is meant to be self.clock or the clock of the element, like in `get`?

File: src/map.rs [hunk 2]
```diff
@@ -344,6 +344,139 @@ impl<K: Ord, V: Val<A> + Default, A: Actor> Map<K, V, A> {
...
+    ///
+    /// let add_ctx = map.read_ctx().derive_add_ctx(actor);
+    /// map.apply(map.update(200, add_ctx, |v, a| v.write("baz", a)));
+    ///
+    ///
+    /// let mut items = Vec::new();
+    ///
+    /// for read_ctx in map.keys() {
+    ///     items.push(*read_ctx.val);
+    /// }
```
Comment (davidrusu) [inline src/map.rs:378]:
```suggestion
    /// let items: Vec<_> = map.keys().map(|key_ctx| *key_ctx.val).collect();
```

File: src/map.rs [hunk 3]
```diff
@@ -344,6 +344,139 @@ impl<K: Ord, V: Val<A> + Default, A: Actor> Map<K, V, A> {
...
+    ///
+    /// let add_ctx = map.read_ctx().derive_add_ctx(actor);
+    /// map.apply(map.update(200, add_ctx, |v, a| v.write("baz", a)));
+    ///
+    ///
+    /// let mut items = Vec::new();
+    ///
+    /// for read_ctx in map.values() {
+    ///     items.push(read_ctx.val.read().val[0]);
+    /// }
```
Comment (davidrusu) [inline src/map.rs:422]: Use `let values = map.values().map(|| ...).collect()` like above

File: src/map.rs [hunk 4]
```diff
@@ -344,6 +344,139 @@ impl<K: Ord, V: Val<A> + Default, A: Actor> Map<K, V, A> {
...
+    /// let add_ctx = map.read_ctx().derive_add_ctx(actor);
+    /// map.apply(map.update(200, add_ctx, |v, a| v.write("baz", a)));
+    ///
+    ///
+    /// let mut items = Vec::new();
+    ///
+    /// for read_ctx in map.iter() {
+    ///     let (key, value) = read_ctx.val;
+    ///     items.push((*key, value.read().val[0]));
+    /// }
```
Comment (davidrusu) [inline src/map.rs:467]: Same here

File: src/map.rs (hunk 1 above)
Comment (davidrusu) [reply src/map.rs]: Yeah you got it right in the next commit, we need the entry clock

**Comment by 5225225** [comment]
Changed the doctests to use iterator methods, and realised we should be returning Iterator and not IntoIterator

**Comment by davidrusu** [comment]
Looks good! Thanks!

**Pull request merged**